frontend/appflowy_flutter/lib/plugins/document/application/cover_image_picker_bloc.dart

```diff
@@ -0,0 +1,181 @@
+
+class CoverImagePickerBloc
+    extends Bloc<CoverImagePickerEvent, CoverImagePickerState> {
+  CoverImagePickerBloc() : super(const CoverImagePickerState.initial()) {
+    on<CoverImagePickerEvent>(
+      (event, emit) async {
+        await event.map(initialEvent: (InitialEvent initialEvent) {
+          emit(const CoverImagePickerState.initial());
+        }, urlSubmit: (UrlSubmit urlSubmit) async {
+          emit(const CoverImagePickerState.loading());
+          final validateImage = await _validateUrl(urlSubmit.path);
+          if (validateImage) {
+            emit(CoverImagePickerState.networkImage(left(urlSubmit.path)));
+          } else {
+            emit(CoverImagePickerState.networkImage(right(FlowyError(
+                msg: LocaleKeys.document_plugins_cover_couldNotFetchImage
+                    .tr()))));
+          }
+        }, pickFileImage: (PickFileImage pickFileImage) async {
+          final imagePickerResults = await _pickImages();
+          if (imagePickerResults != null) {
+            emit(CoverImagePickerState.fileImage(imagePickerResults));
+          } else {
+            emit(const CoverImagePickerState.initial());
+          }
+        }, deleteImage: (DeleteImage deleteImage) {
+          emit(const CoverImagePickerState.initial());
+        }, saveToGallery: (SaveToGallery saveToGallery) async {
+          emit(const CoverImagePickerState.loading());
+          final saveImage = await _saveToGallery(saveToGallery.previousState);
+          if (saveImage != null) {
+            emit(CoverImagePickerState.done(left(saveImage)));
+          } else {
+            emit(CoverImagePickerState.done(
+              right(
+                FlowyError(
+                    msg: LocaleKeys.document_plugins_cover_imageSavingFailed
+                        .tr()),
+              ),
+            ));
+            emit(const CoverImagePickerState.initial());
+          }
+        });
+      },
+    );
+  }
+
+  _saveToGallery(CoverImagePickerState state) async {
+    SharedPreferences prefs = await SharedPreferences.getInstance();
+    List<String> imagePaths = prefs.getStringList(kLocalImagesKey) ?? [];
+    final directory = await _coverPath();
+
+    if (state is FileImagePicked) {
+      try {
+        final path = state.path;
+        final newPath = '$directory/${path.split("\\").last}';
+        final newFile = await File(path).copy(newPath);
+        imagePaths.add(newFile.path);
+        await prefs.setStringList(kLocalImagesKey, imagePaths);
+        return imagePaths;
+      } catch (e) {
+        return null;
+      }
+    } else if (state is NetworkImagePicked) {
+      try {
+        String? url;
+        state.successOrFail.fold((path) {
+          url = path;
+        }, (r) => null);
+        final response = await http.get(Uri.parse(url!));
+
+        final newPath =
+            "$directory/IMG_${Random().nextInt(1000)}.${url!.contains('.png') ? 'png' : url!.contains('.jpeg') ? 'jpeg' : 'jpg'}";
```

Comment:
You should filter the URL that doesn't contains `png`, 'jepg' or `jpg` in `_validateUrl` function. Then you can just use `p.extension(url) ` to get the extension name.

Reply:
And don't use `Random()`, otherwise it will conflict with another image name. You can use `timestamp` to name the image.